tekton/utils/git_reference_manager.go

```diff
@@ -0,0 +1,194 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	tektonv1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+)
+
+// Common branch names that typically cause race conditions when used as mutable references
+var commonBranches = []string{
+	"main",
+	"production",
+	"staging",
+	"development",
+}
+
+// Context key for mock GitHub API base URL (used in tests)
+type contextKey string
+
+const GitHubAPIBaseURLKey contextKey = "github_api_base_url"
+
+// WithGitReferenceRewriting enables git reference rewriting for race condition prevention
+func (b *PipelineRunBuilder) WithGitReferenceRewriting(resolvedSHA, gitURL string) *PipelineRunBuilder {
+	if b.gitRefManager == nil {
+		b.gitRefManager = &GitReferenceManager{
+			EnableAutoRewrite:  false,
+			ValidationWarnings: []string{},
+		}
+	}
+
+	b.gitRefManager.EnableAutoRewrite = true
+	b.gitRefManager.CentralGitRevision = resolvedSHA
+	b.gitRefManager.CentralGitURL = gitURL
+
+	return b
+}
+
+// GetGitReferenceWarnings returns any validation warnings about git references.
+func (b *PipelineRunBuilder) GetGitReferenceWarnings() []string {
+	if b.gitRefManager == nil {
+		return []string{}
+	}
+	return b.gitRefManager.ValidationWarnings
+}
+
+// processGitReferences performs git reference rewriting
+func (b *PipelineRunBuilder) processGitReferences() {
+	if b.gitRefManager == nil || !b.gitRefManager.EnableAutoRewrite || b.pipelineRun.Spec.PipelineSpec == nil {
+		return
+	}
+
+	for i := range b.pipelineRun.Spec.PipelineSpec.Tasks {
+		b.rewriteTaskRef(&b.pipelineRun.Spec.PipelineSpec.Tasks[i])
+	}
+
+	for i := range b.pipelineRun.Spec.PipelineSpec.Finally {
+		b.rewriteTaskRef(&b.pipelineRun.Spec.PipelineSpec.Finally[i])
+	}
+}
+
+// rewriteTaskRef rewrites a single task's git reference
+func (b *PipelineRunBuilder) rewriteTaskRef(task *tektonv1.PipelineTask) {
+	if task.TaskRef == nil || task.TaskRef.ResolverRef.Resolver != "git" {
+		return
+	}
+
+	var url, revision string
+	var revisionIndex = -1
+
+	for i, param := range task.TaskRef.ResolverRef.Params {
+		switch param.Name {
+		case "url":
+			url = param.Value.StringVal
+		case "revision":
+			revision = param.Value.StringVal
+			revisionIndex = i
+		}
+	}
+
+	if revisionIndex >= 0 && shouldRewrite(url, revision, b.gitRefManager.CentralGitURL) {
+		task.TaskRef.ResolverRef.Params[revisionIndex].Value.StringVal = b.gitRefManager.CentralGitRevision
+	}
+}
+
+// shouldRewrite determines if a git reference should be rewritten
+func shouldRewrite(url, revision, centralURL string) bool {
+	urlMatches := strings.Contains(url, centralURL) ||
+		strings.Contains(url, "github.com/konflux-ci/release-service-catalog") ||
```

Comment:
We could move these to a slice e.g. `allowedRepos` and loop them instead? Example:
```go
var allowedRepos = []string{
	"github.com/konflux-ci/release-service-catalog",
	"github.com/konflux-ci/tekton-integration-catalog",
	"github.com/konflux-ci/release-service",
	"github.com/test/repo", // test repo
}

func matchesAllowedUrls(url string) bool {
  for _, allowed := range allowedRepos {
  		if strings.Contains(repoURL, allowed) {
  			return true
  		}
  }
}

// and then use it in shouldRewrite
if !matchesAllowedUrls(url) {
  return false
}
```